File: aisd/aisd/src/main/java/org/example/structure/hashtable/HashTableLineProbing.java
package org.example.structure.hashtable;

import org.example.model.Person;

// Хэш таблица с линейным пробированием
public class HashTableLineProbing {
    private final int size; // Размер хэш-таблицы
    private final Person[] data; // Массив элементов хэш-таблицы
    private final int step; // Шаг для линейного пробирования

    public HashTableLineProbing(int size, int step) {
        this.size = size;
        this.data = new Person[size]; // Инициализация массива элементов
        this.step = step; // Инициализация шага
    }

    public void print() {
        int i = 1;
        System.out.println("======================");
        for (Person element : data) {
            if (element == null || element.isDeleted()) {
                System.out.println("Элемент " + i + " - null"); // Вывод информации о пустых или удаленных элементах
            } else {
                System.out.println("Элемент " + i + " - " + element.getName() + " " + element.getAge()); // Вывод информации о каждом элементе
            }
            i++;
        }
        System.out.println("======================");
    }

    public void insert(Person personModel) {
        int index = getIndex(personModel); // Получение индекса для вставки элемента

        if (data[index] == null || data[index].isDeleted()) {
            data[index] = personModel; // Вставка элемента, если ячейка пуста или содержит удаленный элемент
            return;
        }

        for (int i = 1; index + i * step < size; i++) {
            if (data[index + i * step] == null || data[index + i * step].isDeleted()) {
                data[index + i * step] = personModel; // Вставка элемента на позицию, найденную с использованием линейного пробирования
                return;
            }
        }

        System.out.println("Таблица переполнена"); // Вывод сообщения об ошибке, если таблица переполнена
    }

    public void delete(Person person) {
        int index = getIndex(person); // Получение индекса элемента для удаления

        for (int i = 0; index + i * step < size; i++) {
            // Проверка, что элемент не равен null перед доступом к его свойствам
            if (data[index + i * step] != null &&
                    data[index + i * step].getAge() == person.getAge() &&
                    data[index + i * step].getName().equals(person.getName()) &&
                    data[index + i * step].isDeleted() == person.isDeleted()) {
                data[index + i * step].setDeleted(true); // Установка флага удаления для найденного элемента
            }
        }
    }

    public boolean find(Person personModel) {
        int index = getIndex(personModel); // Получение индекса элемента для поиска

        if (data[index] == null) {
            return false; // Возвращение false, если ячейка пуста
        } else {
            for (int i = 0; index + i * step < size; i++) {
                if (data[index + i * step] == null) {
                    return false; // Возвращение false, если элемент не найден
                }
                if (data[index + i * step].getAge() == personModel.getAge() &&
                        data[index + i * step].getName().equals(personModel.getName()) &&
                        data[index + i * step].isDeleted() == personModel.isDeleted()) {
                    return true; // Возвращение true, если элемент найден
                }
            }
            return false; // Возвращение false, если элемент не найден
        }
    }

    private int getIndex(Person personModel) {
        // Вычисление хэша для определения индекса элемента в массиве
        return (personModel.getAge() * personModel.getAge() +
                personModel.getName().length() * personModel.getName().length() +
                personModel.getAge() * personModel.getName().length()) % size;
    }
}

File: misc/ASD/ASD/HashTableLineProbing.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ASD;
public class HashTableLineProbing
{
    private readonly int _size;
    private readonly PersonModel?[] _data;
    private readonly int _step;
    public HashTableLineProbing(int size, int step)
    {
        _size = size;
        _data = new PersonModel?[_size];
        _step = step;
    }

    public void Print()
    {
        int i = 1;
        foreach (var element in _data)
        {
            if (element == null || element.IsDeleted)
            {
                Console.WriteLine($"Элемент {i} - null");
            }
            else
            {
                Console.WriteLine($"Элемент {i} - {element.Name} {element.Age}");
            }
            i++;
        }
    }

    public void Insert(PersonModel personModel)
    {
        int index = GetIndex(personModel);

        if (_data[index] == null || _data[index].IsDeleted)
        {
            _data[index] = personModel;
            return;
        }

        for (int i = 1; index + i * _step < _size; i++)
        {
            if (_data[index + i * _step] == null || _data[index + i * _step].IsDeleted)
            {
                _data[index + i * _step] = personModel;
                return;
            }
        }

        Console.WriteLine("Таблица переполнена");
    }

    public void Delete(PersonModel personModel)
    {
        int index = GetIndex(personModel);

        for (int i = 0; index + i * _step < _size; i++)
        {
            if (_data[index + i * _step].Age == personModel.Age && _data[index + i * _step].Name == personModel.Name && _data[index + i * _step].IsDeleted == personModel.IsDeleted)
            {
                _data[index + i * _step].IsDeleted = true;
            }
        }
    }

    public bool Find(PersonModel personModel)
    {
        int index = GetIndex(personModel);

        if (_data[index] == null)
        {
            return false;
        }
        else
        {
            for (int i = 0; index + i * _step < _size; i++)
            {
                if (_data[index + i * _step] == null)
                {
                    return false;
                }
                if (_data[index + i * _step].Age == personModel.Age && _data[index + i * _step].Name == personModel.Name && _data[index + i * _step].IsDeleted == personModel.IsDeleted)
                {
                    return true;
                }
            }

            return false;
        }
    }

    private int GetIndex(PersonModel personModel)
        => (personModel.Age * personModel.Age + personModel.Name.Length * personModel.Name.Length + personModel.Age * personModel.Name.Length) % _size;

}
